src/Components/ServiceScene/Breakdowns/LabelValuesBreakdownScene.tsx

```diff
@@ -77,47 +121,151 @@ export class LabelValuesBreakdownScene extends SceneObjectBase<LabelValueBreakdo
-      this.state.$data?.subscribeToState((newState, prevState) => {
-        this.onValuesDataQueryChange(newState, prevState);
+      getFieldsAndMetadataVariable(this).subscribeToState((newState, prevState) => {
+        if (!areArraysEqual(newState.filters, prevState.filters)) {
+          this.runQuery();
+        }
       })
     );
-  }
 
-  private onValuesDataQueryChange(newState: SceneDataState, prevState: SceneDataState) {
-    // Set empty states
-    this.setEmptyStates(newState);
+    // Subscribe to line filter variable changes
+    this._subs.add(
+      getLineFiltersVariable(this).subscribeToState((newState, prevState) => {
+        if (!areArraysEqual(newState.filters, prevState.filters)) {
+          this.runQuery();
+        }
+      })
+    );
 
-    // Set error states
-    this.setErrorStates(newState);
+    // Subscribe to pattern variable changes
+    this._subs.add(
+      getPatternsVariable(this).subscribeToState((newState, prevState) => {
+        if (newState.value !== prevState.value) {
+          this.runQuery();
+        }
+      })
+    );
 
-    // Navigate back to main page if user reduced cardinality to 1
-    this.navigateOnLastFilter(newState);
+    // if this label is detected level
+    if (this.getTagKey() === LEVEL_VARIABLE_VALUE) {
+      // Subscribe to the labels variable
+      this._subs.add(
+        getLabelsVariable(this).subscribeToState(async (newState, prevState) => {
+          if (!areArraysEqual(newState.filters, prevState.filters)) {
+            // Check to see if excluding this label changes the query string before running
+            // If the filter change was for the label we're looking at, there's no need to re-run the query
+            const partialFilterExpression = this.removeValueLabelFromVariableInterpolation();
+            const filterExpression = renderLogQLLabelFilters(newState.filters, [this.getTagKey()]);
+
+            if (partialFilterExpression !== filterExpression) {
+              const queryRunner = this.getSceneQueryRunner();
+              queryRunner?.runQueries();
+            }
+          }
+        })
+      );
+    } else {
+      //otherwise subscribe to levels variable
+      this._subs.add(
+        getLevelsVariable(this).subscribeToState(async (newState, prevState) => {
+          if (!areArraysEqual(newState.filters, prevState.filters)) {
+            // Check to see if excluding this label changes the query string before running
+            // If the filter change was for the label we're looking at, there's no need to re-run the query
+            const partialFilterExpression = this.removeValueLabelFromVariableInterpolation();
+            const filterExpression = renderLevelsFilter(newState.filters, [this.getTagKey()]);
+
+            if (partialFilterExpression !== filterExpression) {
+              const queryRunner = this.getSceneQueryRunner();
+              queryRunner?.runQueries();
+            }
+          }
+        })
+      );
+    }
```

Comment:
First, double checking if these conditionals arent inverted. If the tag key is level, subscribe to labels?

Then, it looks like it could be unified as:

```suggestion
    const variable = this.getTagKey() === LEVEL_VARIABLE_VALUE ? getLabelsVariable(this) : getLevelsVariable(this);
    const renderer = this.getTagKey() === LEVEL_VARIABLE_VALUE ? renderLogQLLabelFilters : renderLevelsFilter;
    this._subs.add(
      variable.subscribeToState(async (newState, prevState) => {
        if (!areArraysEqual(newState.filters, prevState.filters)) {
          // Check to see if excluding this label changes the query string before running
          // If the filter change was for the label we're looking at, there's no need to re-run the query
          const partialFilterExpression = this.removeValueLabelFromVariableInterpolation();
          const filterExpression = renderer(newState.filters, [this.getTagKey()]);

          if (partialFilterExpression !== filterExpression) {
            const queryRunner = this.getSceneQueryRunner();
            queryRunner?.runQueries();
          }
        }
      })
    );
```

Reply:
Yeah good call, much cleaner!

Added in https://github.com/grafana/logs-drilldown/pull/1074/commits/e568e30b6cc24d3c2669a4808e5fbeddfccdbeee